In [1]:
import pandas as pd

In [2]:
df = pd.read_csv("chicago_mini.csv")

In [3]:
df.head()

Start Time             End Time  Trip Duration  \
0  2017-05-29 18:36:27  2017-05-29 18:49:27            780   
1  2017-06-12 19:00:33  2017-06-12 19:24:22           1429   
2  2017-02-13 17:02:02  2017-02-13 17:20:10           1088   
3  2017-04-24 18:39:45  2017-04-24 18:54:59            914   
4  2017-01-26 15:36:07  2017-01-26 15:43:21            434   

                  Start Station                          End Station  \
0     Columbus Dr & Randolph St                 Federal St & Polk St   
1        Kingsbury St & Erie St  Orleans St & Merchandise Mart Plaza   
2         Canal St & Madison St              Paulina Ave & North Ave   
3  Spaulding Ave & Armitage Ave       California Ave & Milwaukee Ave   
4        Clark St & Randolph St         Financial Pl & Congress Pkwy   

    User Type  Gender  Birth Year  
0  Subscriber    Male      1991.0  
1    Customer     NaN         NaN  
2  Subscriber  Female      1982.0  
3  Subscriber    Male      1966.0  
4  Subscriber  Female      1983.0

In [4]:
df.columns

Index(['Start Time', 'End Time', 'Trip Duration', 'Start Station',
       'End Station', 'User Type', 'Gender', 'Birth Year'],
      dtype='object')

In [5]:
df.describe()

Trip Duration   Birth Year
count     400.000000   330.000000
mean      938.215000  1980.533333
std      2466.972184    11.723011
min        97.000000  1949.000000
25%       407.000000  1973.000000
50%       634.000000  1984.000000
75%       997.750000  1989.000000
max     47551.000000  1999.000000

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 400 entries, 0 to 399
Data columns (total 8 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Start Time     400 non-null    object 
 1   End Time       400 non-null    object 
 2   Trip Duration  400 non-null    int64  
 3   Start Station  400 non-null    object 
 4   End Station    400 non-null    object 
 5   User Type      400 non-null    object 
 6   Gender         330 non-null    object 
 7   Birth Year     330 non-null    float64
dtypes: float64(1), int64(1), object(6)
memory usage: 25.1+ KB


In [7]:
df["User Type"].unique()

array(['Subscriber', 'Customer'], dtype=object)

In [8]:
df["Birth Year"].min()

1949.0

In [9]:
df[df["Birth Year"] < 1990]

Start Time             End Time  Trip Duration  \
2    2017-02-13 17:02:02  2017-02-13 17:20:10           1088   
3    2017-04-24 18:39:45  2017-04-24 18:54:59            914   
4    2017-01-26 15:36:07  2017-01-26 15:43:21            434   
5    2017-06-21 09:43:54  2017-06-21 10:01:31           1057   
6    2017-04-24 08:33:31  2017-04-24 08:38:47            316   
..                   ...                  ...            ...   
391  2017-05-21 19:17:39  2017-05-21 19:22:12            273   
393  2017-06-06 05:54:58  2017-06-06 05:59:56            298   
395  2017-02-06 13:34:32  2017-02-06 13:41:35            423   
396  2017-05-22 16:44:08  2017-05-22 16:48:05            237   
398  2017-05-16 08:14:27  2017-05-16 08:28:24            837   

                    Start Station                     End Station   User Type  \
2           Canal St & Madison St         Paulina Ave & North Ave  Subscriber   
3    Spaulding Ave & Armitage Ave  California Ave & Milwaukee Ave  Subscriber   
4          Clark St & Randolph St    Financial Pl & Congress Pkwy  Subscriber   
5            Ogden Ave & Race Ave         Franklin St & Monroe St  Subscriber   
6    Clinton St & Washington Blvd             Morgan St & Lake St  Subscriber   
..                            ...                             ...         ...   
391     Ashland Ave & Division St         Paulina Ave & North Ave  Subscriber   
393      McClurg Ct & Illinois St            State St & Kinzie St  Subscriber   
395      Jefferson St & Monroe St        Racine Ave & Randolph St  Subscriber   
396        Clark St & Randolph St    Clinton St & Washington Blvd  Subscriber   
398          Wabash Ave & 16th St              Daley Center Plaza  Subscriber   

     Gender  Birth Year  
2    Female      1982.0  
3      Male      1966.0  
4    Female      1983.0  
5    Female      1981.0  
6      Male      1986.0  
..      ...         ...  
391    Male      1989.0  
393  Female      1971.0  
395    Male      1973.0  
396    Male      1959.0  
398    Male      1988.0  

[249 rows x 8 columns]

In [10]:
df["Gender"].value_counts()

Male      244
Female     86
Name: Gender, dtype: int64

-------------------

# 1: Most Popular Start Hour

In [11]:
df.head(2)

Start Time             End Time  Trip Duration  \
0  2017-05-29 18:36:27  2017-05-29 18:49:27            780   
1  2017-06-12 19:00:33  2017-06-12 19:24:22           1429   

               Start Station                          End Station   User Type  \
0  Columbus Dr & Randolph St                 Federal St & Polk St  Subscriber   
1     Kingsbury St & Erie St  Orleans St & Merchandise Mart Plaza    Customer   

  Gender  Birth Year  
0   Male      1991.0  
1    NaN         NaN

In [12]:
import pandas as pd

filename = 'chicago_mini.csv'

# load data file into a dataframe
df = pd.read_csv(filename)

# convert the Start Time column to datetime
df['Start Time'] = pd.to_datetime(df["Start Time"])


# extract hour from the Start Time column to create an hour column
df['hour'] = df["Start Time"].dt.hour

# find the most common hour (from 0 to 23)
popular_hour = df["hour"].mode()[0]
    
print('Most Frequent Start Hour:', popular_hour)

Most Frequent Start Hour: 17


In [13]:
#alternative method of getting the mode or checking distribution
count_by_hour = df.groupby("hour")["hour"].count()
count_by_hour

hour
0      4
5      2
6     14
7     23
8     22
9     20
10    16
11    19
12    31
13    26
14    21
15    25
16    29
17    53
18    36
19    26
20    11
21    11
22     7
23     4
Name: hour, dtype: int64

### Conclusion:
As per findings, 17:00 hour is the most popular Starting Hour for biking sharing. The possible reason is people may be off from work around that hour and it may lead to popularlity of using bike sharing services (such as biking to nearest hang out place, or back to home, etc)

------------

# 2: Display a Breakdown of User Types

In [14]:
import pandas as pd

filename = 'chicago_mini.csv'

# load data file into a dataframe
df = pd.read_csv(filename)
df.head()

# print value counts for each user type
user_types = df["User Type"].value_counts()

print(user_types)

Subscriber    330
Customer       70
Name: User Type, dtype: int64


In [15]:
# alternative approach
user_types = df.groupby(["User Type"])["User Type"].count()

print(user_types)

User Type
Customer       70
Subscriber    330
Name: User Type, dtype: int64


### Conculsion:
There are 70 Customer and 330 Subscriber User Type in sample of 400.

------------

# 3: Load and Filter the Dataset

In [28]:
import pandas as pd

CITY_DATA = { 'chicago': 'chicago_mini.csv',
              'new york city': 'new_york_city.csv',
              'washington': 'washington.csv' }

def load_data(city, month, day):
    """
    Loads data for the specified city and filters by month and day if applicable.

    Args:
        (str) city - name of the city to analyze
        (str) month - name of the month to filter by, or "all" to apply no month filter
        (str) day - name of the day of week to filter by, or "all" to apply no day filter
    Returns:
        df - pandas DataFrame containing city data filtered by month and day
    """
    
    # load data file into a dataframe
    df = pd.read_csv(CITY_DATA[city])

    # convert the Start Time column to datetime
    df['Start Time'] = pd.to_datetime(df["Start Time"])

    # extract month and day of week from Start Time to create new columns
    df['month'] = df["Start Time"].dt.month
    df['day_of_week'] = df['Start Time'].dt.day_name ()

    # filter by month if applicable
    if month != 'all':
        pass
        # use the index of the months list to get the corresponding int
        months = ['january', 'february', 'march', 'april', 'may', 'june']
        month = months.index(month) + 1
    
        # filter by month to create the new dataframe
        df = df[df["month"] == month]

    # filter by day of week if applicable
    if day != 'all':
        # filter by day of week to create the new dataframe
        df = df[df["day_of_week"] == day.title()]
    
    return df
    
df = load_data('chicago', 'march', 'friday')

In [29]:
df

Start Time             End Time  Trip Duration  \
40  2017-03-24 13:06:37  2017-03-24 13:10:44            247   
59  2017-03-03 07:55:48  2017-03-03 07:57:41            113   
68  2017-03-17 12:14:50  2017-03-17 12:22:38            468   
75  2017-03-10 13:40:54  2017-03-10 13:45:09            255   
83  2017-03-24 14:15:43  2017-03-24 14:27:04            681   
126 2017-03-24 12:39:19  2017-03-24 12:52:11            772   
224 2017-03-31 19:11:12  2017-03-31 19:18:53            461   
247 2017-03-10 08:21:05  2017-03-10 08:23:28            143   
290 2017-03-24 10:55:53  2017-03-24 11:01:27            334   
343 2017-03-17 17:51:31  2017-03-17 18:00:16            525   
348 2017-03-31 07:47:14  2017-03-31 07:55:38            504   

                   Start Station                End Station   User Type  \
40         Broadway & Berwyn Ave      Clark St & Berwyn Ave  Subscriber   
59        Clark St & Chicago Ave        Wells St & Huron St  Subscriber   
68   Dearborn Pkwy & Delaware Pl     State St & Randolph St  Subscriber   
75            Clark St & Lake St       Rush St & Hubbard St  Subscriber   
83    Sheridan Rd & Lawrence Ave   Broadway & Thorndale Ave  Subscriber   
126        Michigan Ave & Oak St  Cannon Dr & Fullerton Ave  Subscriber   
224      Damen Ave & Cortland St     Damen Ave & Pierce Ave  Subscriber   
247      Damen Ave & Division St  Ashland Ave & Division St  Subscriber   
290    Wacker Dr & Washington St  LaSalle St & Jackson Blvd  Subscriber   
343    Milwaukee Ave & Grand Ave      State St & Pearson St  Subscriber   
348        Canal St & Madison St      Wabash Ave & Adams St  Subscriber   

     Gender  Birth Year  month day_of_week  
40   Female      1961.0      3      Friday  
59     Male      1981.0      3      Friday  
68   Female      1984.0      3      Friday  
75   Female      1983.0      3      Friday  
83     Male      1984.0      3      Friday  
126    Male      1993.0      3      Friday  
224    Male      1989.0      3      Friday  
247    Male      1991.0      3      Friday  
290    Male      1961.0      3      Friday  
343    Male      1989.0      3      Friday  
348    Male      1953.0      3      Friday